Source/WebCore/animation/CSSPropertyAnimation.cpp

```diff
@@ -3910,13 +3911,42 @@ void CSSPropertyAnimation::blendProperties(const CSSPropertyBlendingClient* clie
     }
 }
 
-void CSSPropertyAnimation::blendCustomProperty(const AtomString& customProperty, RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, double progress)
+template <typename T>
+static std::optional<T> typedCustomPropertyValue(const CSSCustomPropertyValue::VariantValue& variantValue)
 {
-    const auto& source = progress < 0.5 ? from : to;
-    if (auto nonInheritedValue = source.nonInheritedCustomProperties().get(customProperty))
-        destination.setNonInheritedCustomPropertyValue(customProperty, CSSCustomPropertyValue::create(*nonInheritedValue));
-    else if (auto inheritedValue = source.inheritedCustomProperties().get(customProperty))
-        destination.setInheritedCustomPropertyValue(customProperty, CSSCustomPropertyValue::create(*inheritedValue));
+    if (std::holds_alternative<CSSCustomPropertyValue::SyntaxValue>(variantValue)) {
+        auto& syntaxValue = std::get<CSSCustomPropertyValue::SyntaxValue>(variantValue);
+        if (std::holds_alternative<T>(syntaxValue))
+            return std::get<T>(syntaxValue);
+    }
+    return std::nullopt;
+}
+
+static Ref<CSSCustomPropertyValue> blendedCSSCustomPropertyValue(const CSSCustomPropertyValue& from, const CSSCustomPropertyValue& to, const CSSPropertyBlendingContext& blendingContext)
+{
+    if (auto possibleFromLength = typedCustomPropertyValue<Length>(from.value())) {
+        ASSERT(typedCustomPropertyValue<Length>(to.value()));
+        auto blendedLength = blendFunc(*possibleFromLength, *typedCustomPropertyValue<Length>(to.value()), blendingContext);
+        return CSSCustomPropertyValue::createForLengthSyntax(from.name(), WTFMove(blendedLength));
+    }
+
+    // Use a discrete interpolation for all other cases.
+    return CSSCustomPropertyValue::create(blendingContext.progress < 0.5 ? from : to);
+}
+
+void CSSPropertyAnimation::blendCustomProperty(const CSSPropertyBlendingClient* client, const AtomString& customProperty, RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, double progress, CompositeOperation compositeOperation, IterationCompositeOperation iterationCompositeOperation, double currentIteration)
+{
+    CSSPropertyBlendingContext blendingContext { progress, false, compositeOperation, client, iterationCompositeOperation, currentIteration };
+
+    auto blendInCustomPropertyValueMap = [&](const CustomPropertyValueMap& (RenderStyle::*map)() const, void (RenderStyle::*setValue)(const AtomString&, Ref<CSSCustomPropertyValue>&&)) {
```

Comment:
Those parameters are nuts. Can you just use `auto` for them? Or not be generic, there are only two cases ever here.